Bugsnag/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSLogger.c

```diff
@@ -147,88 +78,73 @@ bool bsg_kslog_setLogFilename(const char *filename, bool overwrite) {
     }
     int fd = open(filename, openMask, 0644);
     unlikely_if(fd < 0) {
-        writeFmtToLog("KSLogger: Could not open %s: %s", filename,
-                      strerror(errno));
+        bsg_i_kslog_logCBasic("KSLogger: Could not open %s: %s", filename,
+                              strerror(errno));
         return false;
     }
 
     setLogFD(fd);
     return true;
 }
 
-#else // if BSG_KSLogger_CBufferSize <= 0
-
-static FILE *bsg_g_file = NULL;
-
-static inline void setLogFD(FILE *file) {
-    if (g_file != NULL && bsg_g_file != stdout && bsg_g_file != stderr &&
-        bsg_g_file != stdin) {
-        fclose(g_file);
+__printflike(1, 2)
+void bsg_i_kslog_logCBasic(const char *const fmt, ...) {
+    char buf[BSG_KSLOGGER_CBufferSize];
+    const int size = sizeof(buf);
+
+    int len = 0;
+
+    unlikely_if(!fmt) {
+        len = (int)(strncpy(buf + len, "(null)", size - len) - buf);
+    } else if (len < size - 1) {
+        va_list args;
+        va_start(args, fmt);
+        int msglen = vsnprintf(buf + len, size - len, fmt, args);
+        if (msglen > 0) {
+            len += MIN(msglen, size - len);
+        }
+        va_end(args);
     }
-    bsg_g_file = file;
-}
-
-static void writeToLog(const char *const str) {
-    unlikely_if(g_file == NULL) { bsg_g_file = stdout; }
-
-    fprintf(g_file, "%s", str);
-}
-
-static inline void writeFmtArgsToLog(const char *fmt, va_list args) {
-    unlikely_if(g_file == NULL) { bsg_g_file = stdout; }
 
-    if (fmt == NULL) {
-        writeToLog("(null)");
+    if (len < size - 1) {
+        buf[len++] = '\n';
     } else {
-        vfprintf(g_file, fmt, args);
+        buf[size - 1] = '\n';
     }
+
+    write(bsg_g_fd, buf, (size_t)len);
 }
 
-static inline void flushLog(void) { fflush(g_file); }
+__printflike(5, 6)
+void bsg_i_kslog_logC(const char *const level, const char *const file,
+                      const int line, const char *const function,
+                      const char *const fmt, ...) {
+    char buf[BSG_KSLOGGER_CBufferSize];
+    const int size = sizeof(buf);
 
-bool bsg_kslog_setLogFilename(const char *filename, bool overwrite) {
-    if (filename == NULL) {
-        setLogFD(stdout);
-        return true;
+    int len = snprintf(buf, size, "%s: %s:%u: %s(): ",
+                       level, lastPathEntry(file), line, function);
+    if (len < 0) {
+        return;
     }
 
-    FILE *file = fopen(filename, overwrite ? "wb" : "ab");
-    unlikely_if(file == NULL) {
-        writeFmtToLog("KSLogger: Could not open %s: %s", filename,
-                      strerror(errno));
-        return false;
+    unlikely_if(!fmt) {
+        len = (int)(strncpy(buf + len, "(null)", size - len) - buf);
```

Comment:
`strncpy` returns its `dst` parameter, so int((`buf` + `len`) - `buf`) == `len`, which means the later `buf[len++] = '\n'` will replace the `(` in `(null)` with a newline.

I'd feel better if we used less clever code. It took me awhile to puzzle this out ;-)

Reply:
Fair point - I'd feel better if we simply did an early exit if `fmt` is NULL to keep the code simpler - would you be ok with that?

Reply:
Yup that's totally fine. Anything to keep code this low level as simple and non-crash-prone as possible :)